source/MaterialXGenShader/ShaderNode.cpp

```diff
@@ -383,38 +414,78 @@ void ShaderNode::setPaths(const Node& node, const NodeDef& nodeDef, bool include
     }
 }
 
-void ShaderNode::setValues(const Node& node, const NodeDef& nodeDef, GenContext& context)
+void ShaderNode::createMetadata(const NodeDef& nodeDef, GenContext& context)
 {
-    // Copy input values from the given node
-    for (const ValueElementPtr& nodeValue : node.getActiveValueElements())
+    ShaderMetadataRegistryPtr registry = context.getUserData<ShaderMetadataRegistry>(ShaderMetadataRegistry::USER_DATA_NAME);
+    if (!(registry && registry->getAllMetadata().size()))
     {
-        ShaderInput* input = getInput(nodeValue->getName());
-        ValueElementPtr nodeDefInput = nodeDef.getActiveValueElement(nodeValue->getName());
-        if (input && nodeDefInput)
-        {
-            const string& valueString = nodeValue->getResolvedValueString();
-            std::pair<const TypeDesc*, ValuePtr> enumResult;
-            if (context.getShaderGenerator().remapEnumeration(*nodeDefInput, valueString, enumResult))
-            {
-                input->setValue(enumResult.second);
-            }
-            else if (!valueString.empty())
-            {
-                input->setValue(nodeValue->getResolvedValue());
-            }
+        // Early out if no metadata is registered.
+        return;
+    }
 
-            InputPtr inputElem = nodeValue->asA<Input>();
-            if (inputElem)
+    // Set metadata on the node according to the nodedef attributes.
+    ShaderMetadataVecPtr nodeMetadataStorage = getMetadata();
+    for (const string& nodedefAttr : nodeDef.getAttributeNames())
+    {
+        const ShaderMetadata* metadataEntry = registry->findMetadata(nodedefAttr);
+        if (metadataEntry)
+        {
+            const string& attrValue = nodeDef.getAttribute(nodedefAttr);
```

Comment:
Minor: Might be faster to do this test before the registry find() ?

Reply:
The first test is much more likely to fail, which means we will avoid the second search in getAttribute, so I think it's more efficient this way.